[ICE]:  thread 'rustc' (2584) panicked at ... compiler/rustc_codegen_ssa/src/mir/operand.rs:291:18:

Hi, I can try to minimize this in a week or two.

Basically I have forked the sqlite crate and randomly got an ICE in GitHub Actions 🤷‍♀️ So here is the log and all that? The compiler said you'd appreciate a report.

### Code

The workflow triggers from `cargo bench` right here: https://github.com/ReactorScram/sql-peas/blob/15c1114c2a2ee8898c2e2d703e4ac4e184781ede/.github/workflows/build.yml#L29

### Meta

`rustc 1.99.0-nightly (89c61a754 2026-07-23) running on x86_64-unknown-linux-gnu` per logs

### Error output

https://github.com/ReactorScram/sql-peas/actions/runs/30069130681/job/89406009786#step:4:74



```
2026-07-24T05:19:54.6973493Z   nightly-x86_64-unknown-linux-gnu installed - rustc 1.99.0-nightly (89c61a754 2026-07-23)
2026-07-24T05:19:54.6974223Z 
2026-07-24T05:19:54.7066292Z ##[group]Run cargo +nightly bench
2026-07-24T05:19:54.7066619Z [36;1mcargo +nightly bench[0m
2026-07-24T05:19:54.7122710Z shell: /usr/bin/bash -e {0}
2026-07-24T05:19:54.7122977Z ##[endgroup]
2026-07-24T05:19:54.7487441Z     Updating crates.io index
2026-07-24T05:19:56.9395061Z      Locking 15 packages to latest compatible versions
2026-07-24T05:19:56.9430003Z  Downloading crates ...
2026-07-24T05:19:57.2700643Z   Downloaded find-msvc-tools v0.1.9
2026-07-24T05:19:57.2740256Z   Downloaded pin-project-lite v0.2.17
2026-07-24T05:19:57.3237588Z   Downloaded tokio-macros v2.7.1
2026-07-24T05:19:57.3451527Z   Downloaded pkg-config v0.3.33
2026-07-24T05:19:57.3469118Z   Downloaded shlex v2.0.1
2026-07-24T05:19:57.3904769Z   Downloaded quote v1.0.47
2026-07-24T05:19:57.3935326Z   Downloaded sqlite3-sys v0.18.0
2026-07-24T05:19:57.3947289Z   Downloaded unicode-ident v1.0.24
2026-07-24T05:19:57.4004141Z   Downloaded proc-macro2 v1.0.107
2026-07-24T05:19:57.4152703Z   Downloaded cc v1.3.0
2026-07-24T05:19:57.5249380Z   Downloaded syn v2.0.119
2026-07-24T05:19:57.5885456Z   Downloaded tokio v1.53.1
2026-07-24T05:19:57.6754690Z   Downloaded sqlite3-src v0.7.0
2026-07-24T05:19:57.8052455Z   Downloaded temporary v0.7.0
2026-07-24T05:19:57.8588053Z   Downloaded random v0.14.1
2026-07-24T05:19:57.8745724Z    Compiling proc-macro2 v1.0.107
2026-07-24T05:19:57.8746378Z    Compiling shlex v2.0.1
2026-07-24T05:19:57.8754588Z    Compiling find-msvc-tools v0.1.9
2026-07-24T05:19:57.8755253Z    Compiling unicode-ident v1.0.24
2026-07-24T05:19:57.9356318Z    Compiling pkg-config v0.3.33
2026-07-24T05:19:57.9845191Z    Compiling cc v1.3.0
2026-07-24T05:19:58.0104154Z    Compiling quote v1.0.47
2026-07-24T05:19:58.2489519Z    Compiling pin-project-lite v0.2.17
2026-07-24T05:19:58.2850208Z    Compiling random v0.14.1
2026-07-24T05:19:58.3143853Z    Compiling temporary v0.7.0
2026-07-24T05:19:58.6945164Z    Compiling syn v2.0.119
2026-07-24T05:19:58.8062551Z    Compiling sqlite3-src v0.7.0
2026-07-24T05:19:58.9288056Z    Compiling sqlite3-sys v0.18.0
2026-07-24T05:19:58.9991820Z    Compiling sql-peas v0.37.0 (/home/runner/work/sql-peas/sql-peas)
2026-07-24T05:20:00.4018236Z    Compiling tokio-macros v2.7.1
2026-07-24T05:20:00.6175454Z    Compiling tokio v1.53.1
2026-07-24T05:20:01.6307821Z error: internal compiler error: /rustc-dev/89c61a7545da48b06116675b888398d02a4064c7/compiler/rustc_codegen_ssa/src/mir/operand.rs:291:18: not immediate: OperandRef(Uninit @ TyAndLayout { ty: unsafe_cell::UnsafeCell<MaybeUninit<runtime::task::Notified<Arc<multi_thread::handle::Handle>>>>, layout: Layout { size: Size(8 bytes), align: AbiAlign { abi: Align(8 bytes) }, backend_repr: Scalar(union pointer), fields: Arbitrary { offsets: [Size(0 bytes)], in_memory_order: [0] }, largest_niche: None, uninhabited: false, variants: Single { index: 0 }, max_repr_align: None, unadjusted_abi_align: Align(8 bytes), randomization_seed: 13045912409502673597 } })
2026-07-24T05:20:01.6312502Z 
2026-07-24T05:20:01.6312937Z 
2026-07-24T05:20:01.6314195Z thread 'rustc' (2584) panicked at /rustc-dev/89c61a7545da48b06116675b888398d02a4064c7/compiler/rustc_codegen_ssa/src/mir/operand.rs:291:18:
2026-07-24T05:20:01.6315539Z Box<dyn Any>
2026-07-24T05:20:01.6316040Z stack backtrace:
2026-07-24T05:20:01.6351789Z    0:     0x7fe461421b06 - <<std[dfd1104ffe190864]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[ed3f5ec3e655ea61]::fmt::Display>::fmt
2026-07-24T05:20:01.6356118Z    1:     0x7fe461a0dd52 - core[ed3f5ec3e655ea61]::fmt::write
2026-07-24T05:20:01.6357669Z    2:     0x7fe461436c5c - <std[dfd1104ffe190864]::sys::stdio::unix::Stderr as core[ed3f5ec3e655ea61]::io::write::Write>::write_fmt
2026-07-24T05:20:01.6359034Z    3:     0x7fe4613f373a - std[dfd1104ffe190864]::panicking::default_hook::{closure#0}
2026-07-24T05:20:01.6360213Z    4:     0x7fe4614141a3 - std[dfd1104ffe190864]::panicking::default_hook
2026-07-24T05:20:01.6362140Z    5:     0x7fe46039684b - std[dfd1104ffe190864]::panicking::update_hook::<alloc[1c8dbdf3d088588b]::boxed::Box<rustc_driver_impl[6b51d1876d1cc845]::install_ice_hook::{closure#1}>>::{closure#0}
2026-07-24T05:20:01.6364208Z    6:     0x7fe461414642 - std[dfd1104ffe190864]::panicking::panic_with_hook
2026-07-24T05:20:01.6365576Z    7:     0x7fe4603c3df1 - std[dfd1104ffe190864]::panicking::begin_panic::<rustc_errors[26ceb3296c1369aa]::ExplicitBug>::{closure#0}
2026-07-24T05:20:01.6367745Z    8:     0x7fe4603bcdb6 - std[dfd1104ffe190864]::sys::backtrace::__rust_end_short_backtrace::<std[dfd1104ffe190864]::panicking::begin_panic<rustc_errors[26ceb3296c1369aa]::ExplicitBug>::{closure#0}, !>
2026-07-24T05:20:01.6369718Z    9:     0x7fe4603bab77 - std[dfd1104ffe190864]::panicking::begin_panic::<rustc_errors[26ceb3296c1369aa]::ExplicitBug>
2026-07-24T05:20:01.6371662Z   10:     0x7fe4603cf9b1 - <rustc_errors[26ceb3296c1369aa]::diagnostic::BugAbort as rustc_errors[26ceb3296c1369aa]::diagnostic::EmissionGuarantee>::emit_producing_guarantee
2026-07-24T05:20:01.6373662Z   11:     0x7fe46096aeb9 - rustc_middle[4ac8260c26aa2f27]::util::bug::opt_span_bug_fmt::<rustc_span[e8aa067e7a22a24]::span_encoding::Span>::{closure#0}
2026-07-24T05:20:01.6376245Z   12:     0x7fe46096b022 - rustc_middle[4ac8260c26aa2f27]::ty::context::tls::with_opt::<rustc_middle[4ac8260c26aa2f27]::util::bug::opt_span_bug_fmt<rustc_span[e8aa067e7a22a24]::span_encoding::Span>::{closure#0}, !>::{closure#0}
2026-07-24T05:20:01.6379459Z   13:     0x7fe46095582b - rustc_middle[4ac8260c26aa2f27]::ty::context::tls::with_context_opt::<rustc_middle[4ac8260c26aa2f27]::ty::context::tls::with_opt<rustc_middle[4ac8260c26aa2f27]::util::bug::opt_span_bug_fmt<rustc_span[e8aa067e7a22a24]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
2026-07-24T05:20:01.6381716Z   14:     0x7fe45d9e2f64 - rustc_middle[4ac8260c26aa2f27]::util::bug::bug_fmt
2026-07-24T05:20:01.6383469Z   15:     0x7fe462e53f4a - rustc_codegen_ssa[4c42488230787b2]::mir::codegen_mir::<rustc_codegen_llvm[864e77ff6577e18b]::builder::GenericBuilder<rustc_codegen_llvm[864e77ff6577e18b]::context::FullCx>>
2026-07-24T05:20:01.6385806Z   16:     0x7fe461ac40d9 - rustc_codegen_llvm[864e77ff6577e18b]::base::compile_codegen_unit::module_codegen
2026-07-24T05:20:01.6387670Z   17:     0x7fe462d44f55 - <rustc_codegen_llvm[864e77ff6577e18b]::LlvmCodegenBackend as rustc_codegen_ssa[4c42488230787b2]::traits::backend::ExtraBackendMethods>::compile_codegen_unit
2026-07-24T05:20:01.6389984Z   18:     0x7fe462d413f2 - rustc_codegen_ssa[4c42488230787b2]::base::codegen_crate::<rustc_codegen_llvm[864e77ff6577e18b]::LlvmCodegenBackend, rustc_codegen_llvm[864e77ff6577e18b]::ModuleLlvm>
2026-07-24T05:20:01.6392244Z   19:     0x7fe462d3f60d - <rustc_codegen_llvm[864e77ff6577e18b]::LlvmCodegenBackend as rustc_codegen_ssa[4c42488230787b2]::traits::backend::CodegenBackend>::codegen_crate
2026-07-24T05:20:01.6394190Z   20:     0x7fe462d4a6e0 - <rustc_interface[1cd944d9e5759701]::queries::Linker>::codegen_and_build_linker
2026-07-24T05:20:01.6396276Z   21:     0x7fe462badf10 - rustc_interface[1cd944d9e5759701]::interface::run_compiler::<(), rustc_driver_impl[6b51d1876d1cc845]::run_compiler::{closure#0}>::{closure#2}
2026-07-24T05:20:01.6399867Z   22:     0x7fe462c17424 - std[dfd1104ffe190864]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[1cd944d9e5759701]::util::run_in_thread_with_globals<rustc_interface[1cd944d9e5759701]::util::run_in_thread_pool_with_globals<rustc_interface[1cd944d9e5759701]::interface::run_compiler<(), rustc_driver_impl[6b51d1876d1cc845]::run_compiler::{closure#0}>::{closure#2}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
2026-07-24T05:20:01.6407142Z   23:     0x7fe462c171ed - <std[dfd1104ffe190864]::thread::lifecycle::spawn_unchecked<rustc_interface[1cd944d9e5759701]::util::run_in_thread_with_globals<rustc_interface[1cd944d9e5759701]::util::run_in_thread_pool_with_globals<rustc_interface[1cd944d9e5759701]::interface::run_compiler<(), rustc_driver_impl[6b51d1876d1cc845]::run_compiler::{closure#0}>::{closure#2}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[ed3f5ec3e655ea61]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
2026-07-24T05:20:01.6410875Z   24:     0x7fe462c23406 - <std[dfd1104ffe190864]::sys::thread::unix::Thread>::new::thread_start
2026-07-24T05:20:01.6411818Z   25:     0x7fe45c29caa4 - <unknown>
2026-07-24T05:20:01.6412414Z   26:     0x7fe45c329c6c - <unknown>
2026-07-24T05:20:01.6412840Z   27:                0x0 - <unknown>
2026-07-24T05:20:01.6413110Z 
2026-07-24T05:20:01.6413856Z note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md
2026-07-24T05:20:01.6415102Z 
2026-07-24T05:20:01.6415404Z note: please make sure that you have updated to the latest nightly
2026-07-24T05:20:01.6415934Z 
2026-07-24T05:20:01.6416855Z note: please attach the file at `/home/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.53.1/rustc-ice-2026-07-24T05_20_00-2582.txt` to your bug report
2026-07-24T05:20:01.6417931Z 
2026-07-24T05:20:01.6418308Z note: rustc 1.99.0-nightly (89c61a754 2026-07-23) running on x86_64-unknown-linux-gnu
2026-07-24T05:20:01.6418771Z 
2026-07-24T05:20:01.6419050Z note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no -C strip=debuginfo
2026-07-24T05:20:01.6419412Z 
2026-07-24T05:20:01.6419572Z note: some of the compiler flags provided by cargo are hidden
2026-07-24T05:20:01.6419810Z 
2026-07-24T05:20:01.6419910Z query stack during panic:
2026-07-24T05:20:01.6420134Z end of query stack
2026-07-24T05:20:01.7022748Z error: could not compile `tokio` (lib)
2026-07-24T05:20:01.7023161Z 
2026-07-24T05:20:01.7023296Z Caused by:
2026-07-24T05:20:01.7029620Z   process didn't exit successfully: `/home/runner/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/rustc --crate-name tokio --edition=2021 /home/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/tokio-1.53.1/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C embed-bitcode=no --cfg 'feature="default"' --cfg 'feature="macros"' --cfg 'feature="rt"' --cfg 'feature="rt-multi-thread"' --cfg 'feature="tokio-macros"' --check-cfg 'cfg(docsrs,test)' --check-cfg 'cfg(feature, values("bytes", "default", "fs", "full", "io-std", "io-uring", "io-util", "libc", "macros", "mio", "net", "parking_lot", "process", "rt", "rt-multi-thread", "schedule-latency", "signal", "signal-hook-registry", "socket2", "sync", "taskdump", "test-util", "time", "tokio-macros", "tracing", "windows-sys"))' -C metadata=a8cf4d964337c1fc -C extra-filename=-1690fb908584d4ff --out-dir /home/runner/work/sql-peas/sql-peas/target/release/deps -C strip=debuginfo -L dependency=/home/runner/work/sql-peas/sql-peas/target/release/deps --extern pin_project_lite=/home/runner/work/sql-peas/sql-peas/target/release/deps/libpin_project_lite-c7fff26b062085d8.rmeta --extern tokio_macros=/home/runner/work/sql-peas/sql-peas/target/release/deps/libtokio_macros-825aed7222b7ad95.so --cap-lints allow` (exit status: 101)
2026-07-24T05:20:01.7118252Z ##[error]Process completed with exit code 101.
```

--- a/compiler/rustc_codegen_ssa/src/mir/block.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/block.rs
@@ -580,10 +580,10 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
 
             PassMode::Direct(_) | PassMode::Pair(..) => {
                 let op = self.codegen_consume(bx, mir::Place::return_place().as_ref());
-                if let Ref(place_val) = op.val {
-                    bx.load_from_place(bx.backend_type(op.layout), place_val)
-                } else {
-                    op.immediate_or_packed_pair(bx)
+                match op.val {
+                    Ref(place_val) => bx.load_from_place(bx.backend_type(op.layout), place_val),
+                    Uninit => bx.cx().const_undef(bx.cx().immediate_backend_type(op.layout)),
+                    _ => op.immediate_or_packed_pair(bx),
                 }
             }
 

--- a/tests/codegen-llvm/uninit-return-value.rs
+++ b/tests/codegen-llvm/uninit-return-value.rs
@@ -1,0 +1,13 @@
+// Regression test for https://github.com/rust-lang/rust/issues/159815
+
+#![crate_type = "lib"]
+
+use std::mem::MaybeUninit;
+
+// CHECK-LABEL: @f
+#[no_mangle]
+pub fn f() -> MaybeUninit<*const ()> {
+    // CHECK: start:
+    // CHECK-NEXT: ret ptr undef
+    const { MaybeUninit::uninit() }
+}

